### DISCUSSION TITLE
Add new network to py-substrate-interface

### DISCUSSION THREAD

**Discussion by yeousunn** [polkascan/py-substrate-interface#286] [Q&A] [answered]
Hi, 
I am required to generate an offline extrinsic payload for our network, I am trying to use the py-substrate-interface, for normal query I am able to get the information, however, when I try to generate an extrinsic I am getting error message "Transaction has a bad signature". I think this has something to do with the `type_registry_preset` in the following:-
`
substrate = SubstrateInterface(url=host, 
    ss58_format=42, 
    type_registry_preset="development"
)
`

Since the network I am working on is not registered, in the `type_registry_preset` I am setting as "development". 
I read that the types needs to be registered or I need to set the types manually. I am trying to find how to do it, with no success yet.

Can anyone direct me to the right documents?

**Edit:-** 
I found that the registry is stored in https://raw.githubusercontent.com/polkascan/py-scale-codec/v1.0/scalecodec/type_registry/ and locally in the repository [py-scale-codec](https://github.com/polkascan/py-scale-codec/tree/master/scalecodec/type_registry). 

How do I generate this type registry for my network?

Thank you,
Yeou Sunn

**Comment by arjanz** [comment, marked as answer]
Probably you don't need to specify a `type_registry_preset`, if your network utilises `MetadataV14`. Before V14, type definitions were required to be administered manually, as described here: https://github.com/polkascan/py-scale-codec/issues/94#issuecomment-1346205326

If the runtime is pre-V14 or there are some exotic changes to the default Substrate implementation, you will need some custom type definitions (or overrides). First you can check if a PolkadotJS type registry JSON file exists, for example for Polymesh there is one maintained here: https://github.com/PolymeshAssociation/polymesh-types/blob/main/polymesh_schema.json

Then you convert the format to match the py-scale-codec format (most different is format for `Enum` and `Struct` types) and then you add that JSON file as follows:

```python
custom_type_registry = load_type_registry_file("my-custom-types.json")

substrate = SubstrateInterface(
    url="ws://127.0.0.1:9944",
    ss58_format=42,
    type_registry=custom_type_registry
)
 
```

Example file for Moonbeam (exceptions are address and signature definitions):
```json
{
  "runtime_id": 900,
  "types": {
    "EthereumAddress": "H160",
    "Address": "EthereumAddress",
    "LookupSource": "EthereumAddress",
    "AccountId": "EthereumAddress",
    "ExtrinsicSignature": "EcdsaSignature"
  }
}
```

Example for Acala (exception is different `MultiSignature` definition):
```json
{
  "types": {
    "MultiSignature": {
      "type": "enum",
      "type_mapping": [
        [
          "Ed25519",
          "Ed25519Signature"
        ],
        [
          "Sr25519",
          "Sr25519Signature"
        ],
        [
          "Ecdsa",
          "EcdsaSignature"
        ],
        [
          "Ethereum",
          "[u8; 65]"
        ],
        [
          "Eip712",
          "[u8; 65]"
        ]
      ]
    }
  }
}
```

**Reply by yeousunn** [reply]
@arjanz Thank you for the reply, 
This helps. I will try it out.
Thank you
